cmd/av/stack_sync.go

```diff
@@ -152,64 +256,61 @@ base branch.
 		}
 
 		_, _ = fmt.Fprintf(os.Stderr, "Stack sync complete\n")
+
+		// Return to the starting branch when we're done
+		if _, err := repo.CheckoutBranch(&git.CheckoutBranch{
+			Name: state.OriginalBranch,
+		}); err != nil {
+			return err
+		}
+
 		return nil
 	},
 }
 
-// doStackSync performs the stack sync operation.
-// It returns a boolean indicating whether or not the sync was interrupted
-// (if true, the stack sync can be resumed with the --continue flag).
-// TODO[mvp]:
-//    This whole function is kind of a hot mess, and a lot of it is rooted in
-//    how awkward the API is for dealing with "branch trees" (terminology is
-//    hard when you're creating a meta-tree-structure on top of git branches).
-//    Let's refactor that a bit and then re-work this.
-func doStackSync(repo *git.Repo, config stackSyncConfig) (bool, error) {
-	branches, err := meta.ReadAllBranches(repo)
-	if err != nil {
-		return false, err
-	}
+type syncResult struct {
+	Conflict      bool
+	CurrentBranch string
+}
 
-	currentBranch, err := repo.CurrentBranchName()
+// Find all the ancestor branches of the given branch name and append them to
+// the given slice (in topological order: a comes before b if a is an ancestor
+// of b).
+func previousBranches(branches map[string]meta.Branch, name string) ([]string, error) {
+	current, ok := branches[name]
+	if !ok {
+		return nil, errors.Errorf("branch metadata not found for %q", name)
+	}
+	if current.Parent == "" {
+		return nil, nil
+	}
+	previous, err := previousBranches(branches, current.Parent)
 	if err != nil {
-		return false, err
+		return nil, err
 	}
+	return append(previous, current.Parent), nil
+}
 
-	for {
-		currentMeta, ok := branches[currentBranch]
-		if !ok {
-			return false, errors.Errorf("stack metadata not found for branch %q", currentBranch)
-		}
-		res, err := stacks.SyncBranch(repo, &stacks.SyncBranchOpts{
-			Parent:   currentMeta.Parent,
-			Continue: config.Continue,
-		})
+func subsequentBranches(branches map[string]meta.Branch, name string) ([]string, error) {
+	logrus.Debugf("finding subsequent branches for %q", name)
+	var res []string
+	branchMeta, ok := branches[name]
+	if !ok {
```

Comment:
I've lost track of this now, so pardon me the dump question. When do we store the branch meta? How do we guarantee that branchMeta always exist at this point?

Reply:
We store it when you create the branch with `av stack branch`. It should always exist here because we always write the metadata for a branch and its parent when creating a new branch with `av stack branch`.